// @mui/x-charts/LineChart/useLinePlotData.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLinePlotData = useLinePlotData;
var React = _interopRequireWildcard(require("react"));
var _warning = require("@mui/x-internals/warning");
var _d3Shape = require("@mui/x-charts-vendor/d3-shape");
var _useChartGradientId = require("../hooks/useChartGradientId");
var _scaleGuards = require("../internals/scaleGuards");
var _sampling = require("../internals/plugins/featurePlugins/useChartCartesianAxis/sampling.selectors");
var _useChartCartesianAxisRendering = require("../internals/plugins/featurePlugins/useChartCartesianAxis/useChartCartesianAxisRendering.selectors");
var _useChartSeriesConfig = require("../internals/plugins/corePlugins/useChartSeriesConfig");
var _useStore = require("../internals/store/useStore");
var _getCurve = require("../internals/getCurve");
var _hooks = require("../hooks");
var _constants = require("../constants");
function useLinePlotData(xAxes, yAxes,
/**
 * Overrides the zoom span and pixel size used for sampling. The zoom-slider preview passes its
 * full-range zoom and own width so its density stays stable while the main chart zooms, instead
 * of reading the active zoom from the store.
 */
samplingOverride) {
  const seriesData = (0, _hooks.useLineSeriesContext)();
  const defaultXAxisId = (0, _hooks.useXAxes)().xAxisIds[0];
  const defaultYAxisId = (0, _hooks.useYAxes)().yAxisIds[0];
  const getGradientId = (0, _useChartGradientId.useChartGradientIdBuilder)();
  const drawingArea = (0, _hooks.useDrawingArea)();
  const store = (0, _useStore.useStore)();
  const samplingState = store.use(_sampling.selectorChartSamplingState);
  const sampledSeries = store.use(_sampling.selectorChartSamplingPyramids);
  const activeZoomMap = store.use(_useChartCartesianAxisRendering.selectorChartZoomMap);
  const zoomOptions = store.use(_useChartCartesianAxisRendering.selectorChartZoomOptionsLookup);
  const sampler = store.use(_useChartSeriesConfig.selectorChartSeriesConfig).line?.sampler;
  const zoomMap = samplingOverride?.zoomMap ?? activeZoomMap;
  const samplingSize = samplingOverride?.availableSize ?? drawingArea.width;

  // Skip the line animation while sampling is on, plus the first render after it turns off: the
  // point count changes, so the path would morph.
  const lineMethod = samplingState?.methods.line;
  const samplingEnabled = lineMethod != null && lineMethod !== 'none';
  const wasSamplingEnabled = React.useRef(samplingEnabled);
  const skipSamplingAnimation = samplingEnabled || wasSamplingEnabled.current;
  React.useEffect(() => {
    wasSamplingEnabled.current = samplingEnabled;
  }, [samplingEnabled]);

  // This memo prevents odd line chart behavior when hydrating.
  const allData = React.useMemo(() => {
    if (seriesData === undefined) {
      return [];
    }
    const {
      series,
      stackingGroups
    } = seriesData;
    const linePlotData = [];
    for (const stackingGroup of stackingGroups) {
      const groupIds = stackingGroup.ids;
      for (const seriesId of groupIds) {
        const {
          xAxisId = defaultXAxisId,
          yAxisId = defaultYAxisId,
          stackedData,
          visibleStackedData,
          data,
          connectNulls,
          curve,
          strictStepCurve
        } = series[seriesId];
        if (!(xAxisId in xAxes) || !(yAxisId in yAxes)) {
          continue;
        }
        const xScale = xAxes[xAxisId].scale;
        const xPosition = (0, _hooks.getValueToPositionMapper)(xScale);
        const yScale = yAxes[yAxisId].scale;
        const xData = xAxes[xAxisId].data;
        const gradientId = yAxes[yAxisId].colorScale && getGradientId(yAxisId) || xAxes[xAxisId].colorScale && getGradientId(xAxisId) || undefined;
        if (process.env.NODE_ENV !== 'production') {
          if (xData === undefined) {
            // TODO: fix mui/no-guarded-throw
            // eslint-disable-next-line mui/no-guarded-throw
            throw new Error(`MUI X Charts: ${xAxisId === _constants.DEFAULT_X_AXIS_KEY ? 'The first `xAxis`' : `The x-axis with id "${xAxisId}"`} should have a data property to be able to display a line plot. ` + 'The x-axis data defines the positions for each point in the line. ' + 'Provide a data array to the x-axis configuration.');
          }
          if (xData.length < stackedData.length) {
            (0, _warning.warnOnce)(`MUI X Charts: The data length of the x axis (${xData.length} items) is lower than the length of series (${stackedData.length} items).`, 'error');
          }
        }
        const shouldExpand = curve?.includes('step') && !strictStepCurve && (0, _scaleGuards.isOrdinalScale)(xScale);

        // Only the plain (non-step) path is sampled; step expansion falls back to the full render.
        const built = sampledSeries[seriesId];
        const zoom = zoomMap?.get(xAxisId) ?? (built ? (0, _sampling.getFullRangeZoom)(xAxisId) : undefined);
        let sampledBuckets = null;
        if (samplingEnabled && built && zoom && !shouldExpand && xData) {
          // Pro owns the sampling math; community flattens the buckets into a polyline.
          sampledBuckets = sampler?.sample?.({
            built,
            zoom,
            availableSize: samplingSize,
            minSpan: zoomOptions[xAxisId]?.minSpan ?? 0,
            algorithm: lineMethod,
            getValues: () => Float64Array.from(visibleStackedData, point => point[1])
          }) ?? null;
        }
        let formattedData;
        if (sampledBuckets) {
          // Buckets already hold the indices to render; flatten them.
          formattedData = sampledBuckets.flatMap(bucket => Array.from(bucket.indices, index => ({
            x: xData[index],
            y: visibleStackedData[index],
            nullData: data[index] == null
          })));
        } else {
          formattedData = xData?.flatMap((x, index) => {
            const nullData = data[index] == null;
            if (shouldExpand) {
              const rep = [{
                x,
                y: visibleStackedData[index],
                nullData,
                isExtension: false
              }];
              if (!nullData && (index === 0 || data[index - 1] == null)) {
                rep.unshift({
                  x: (xScale(x) ?? 0) - (xScale.step() - xScale.bandwidth()) / 2,
                  y: visibleStackedData[index],
                  nullData,
                  isExtension: true
                });
              }
              if (!nullData && (index === data.length - 1 || data[index + 1] == null)) {
                rep.push({
                  x: (xScale(x) ?? 0) + (xScale.step() + xScale.bandwidth()) / 2,
                  y: visibleStackedData[index],
                  nullData,
                  isExtension: true
                });
              }
              return rep;
            }
            return {
              x,
              y: visibleStackedData[index],
              nullData
            };
          }) ?? [];
        }
        const d3Data = connectNulls ? formattedData.filter(d => !d.nullData) : formattedData;
        const hidden = series[seriesId].hidden;
        const linePath = (0, _d3Shape.line)().x(d => d.isExtension ? d.x : xPosition(d.x)).defined(d => connectNulls || !d.nullData || !!d.isExtension).y(d => {
          if (hidden) {
            return yScale(d.y[0]);
          }
          return yScale(d.y[1]);
        });
        const d = linePath.curve((0, _getCurve.getCurveFactory)(curve))(d3Data) || '';
        linePlotData.push({
          color: series[seriesId].color,
          gradientId,
          d,
          seriesId,
          hidden: series[seriesId].hidden,
          isSampled: skipSamplingAnimation
        });
      }
    }
    return linePlotData;
  }, [seriesData, defaultXAxisId, defaultYAxisId, xAxes, yAxes, getGradientId, samplingEnabled, lineMethod, sampledSeries, zoomMap, samplingSize, zoomOptions, sampler, skipSamplingAnimation]);
  return allData;
}
