// @mui/x-charts/internals/plugins/featurePlugins/useChartCartesianAxis/sampling.selectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFullRangeZoom = getFullRangeZoom;
exports.selectorChartSamplingState = exports.selectorChartSamplingPyramids = exports.selectorChartHighlightBucketSize = void 0;
var _store = require("@mui/x-internals/store");
var _useChartSeries = require("../../corePlugins/useChartSeries");
var _useChartSeriesConfig = require("../../corePlugins/useChartSeriesConfig");
var _useChartDimensions = require("../../corePlugins/useChartDimensions");
var _useChartCartesianAxisRendering = require("./useChartCartesianAxisRendering.selectors");
const EMPTY_PYRAMIDS = {};
const EMPTY_BUCKET_SIZES = new Map();

/**
 * Full-range zoom (span 100%) for a non-zoomable axis. Sampling gates on a zoom entry; a static axis
 * has none, so we sample it at full range — the `MAX_RENDERED_POINTS` cap still applies. Mirrors the
 * full-range zoom the zoom-slider preview uses to keep its density stable.
 */
function getFullRangeZoom(axisId) {
  return {
    axisId,
    start: 0,
    end: 100
  };
}
const selectorChartSamplingState = state => state.sampling;

/**
 * Built sampling structures keyed by series id. Memoized, so rebuilt only on data change. Builds
 * only for series types that are enabled and register a `sampler`; the plot hook reads them back.
 */
exports.selectorChartSamplingState = selectorChartSamplingState;
const selectorChartSamplingPyramids = exports.selectorChartSamplingPyramids = (0, _store.createSelectorMemoized)(selectorChartSamplingState, _useChartSeries.selectorChartSeriesProcessed, _useChartSeriesConfig.selectorChartSeriesConfig, function selectorChartSamplingPyramids(samplingState, processedSeries, seriesConfig) {
  if (!samplingState?.enabled) {
    return EMPTY_PYRAMIDS;
  }
  let sampled;
  Object.keys(processedSeries).forEach(seriesType => {
    const method = samplingState.methods[seriesType];
    if (!method || method === 'none') {
      return;
    }
    const sampler = seriesConfig[seriesType]?.sampler;
    const typeData = processedSeries[seriesType];
    if (!sampler || !typeData) {
      return;
    }
    for (const seriesId of typeData.seriesOrder) {
      const built = sampler.build(typeData.series[seriesId]);
      if (built) {
        sampled ?? (sampled = {});
        sampled[seriesId] = built;
      }
    }
  });
  return sampled ?? EMPTY_PYRAMIDS;
});

/**
 * Merged bucket size per band axis, to widen the axis highlight over a sampled bucket. Empty when
 * sampling is off or no sampler is registered. The math lives in pro; community reads the integers.
 */
const selectorChartHighlightBucketSize = exports.selectorChartHighlightBucketSize = (0, _store.createSelectorMemoized)(selectorChartSamplingState, _useChartCartesianAxisRendering.selectorChartZoomMap, _useChartCartesianAxisRendering.selectorChartXAxis, _useChartCartesianAxisRendering.selectorChartYAxis, _useChartDimensions.selectorChartDrawingArea, _useChartCartesianAxisRendering.selectorChartZoomOptionsLookup, _useChartSeriesConfig.selectorChartSeriesConfig, function selectorChartHighlightBucketSize(samplingState, zoomMap, xAxis, yAxis, drawingArea, zoomOptions, seriesConfig) {
  if (!samplingState?.enabled) {
    return EMPTY_BUCKET_SIZES;
  }
  const bucketSizeAt = seriesConfig.bar?.sampler?.bucketSizeAt ?? seriesConfig.line?.sampler?.bucketSizeAt;
  if (!bucketSizeAt) {
    return EMPTY_BUCKET_SIZES;
  }
  const bucketSizes = new Map();
  [[xAxis, drawingArea.width], [yAxis, drawingArea.height]].forEach(([axes, availableSize]) => {
    axes.axisIds.forEach(axisId => {
      const data = axes.axis[axisId].data;
      const zoom = zoomMap?.get(axisId) ?? getFullRangeZoom(axisId);
      const minSpan = zoomOptions[axisId]?.minSpan ?? 0;
      if (data) {
        bucketSizes.set(axisId, bucketSizeAt(zoom.end - zoom.start, {
          dataLength: data.length,
          availableSize,
          minSpan
        }));
      }
    });
  });
  return bucketSizes;
});
