// @mui/x-charts/PieChart/seriesConfig/index.mjs
import seriesProcessor from "./seriesProcessor.mjs";
import getColor from "./getColor.mjs";
import legendGetter from "./legend.mjs";
import tooltipGetter from "./tooltip.mjs";
import seriesLayout from "./seriesLayout.mjs";
import getSeriesWithDefaultValues from "./getSeriesWithDefaultValues.mjs";
import { selectorTooltipItemPosition } from "./tooltipPosition.mjs";
import keyboardFocusHandler from "./keyboardFocusHandler.mjs";
import { identifierSerializerSeriesIdDataIndex } from "../../internals/identifierSerializer.mjs";
import { identifierCleanerSeriesIdDataIndex } from "../../internals/identifierCleaner.mjs";
import { createIsHighlighted, createIsFaded } from "../../internals/plugins/featurePlugins/useChartHighlight/index.mjs";
import descriptionGetter from "./descriptionGetter.mjs";
export const pieSeriesConfig = {
  colorProcessor: getColor,
  seriesProcessor,
  seriesLayout,
  legendGetter,
  tooltipGetter,
  selectorTooltipItemPosition,
  getSeriesWithDefaultValues,
  keyboardFocusHandler,
  identifierSerializer: identifierSerializerSeriesIdDataIndex,
  identifierCleaner: identifierCleanerSeriesIdDataIndex,
  descriptionGetter,
  isHighlightedCreator: createIsHighlighted,
  isFadedCreator: createIsFaded
};
