// @mui/x-charts/ChartsAxisHighlight/ChartsXAxisHighlight.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ChartsXHighlight;
var React = _interopRequireWildcard(require("react"));
var _getValueToPositionMapper = require("../hooks/getValueToPositionMapper");
var _scaleGuards = require("../internals/scaleGuards");
var _useStore = require("../internals/store/useStore");
var _useChartCartesianAxis = require("../internals/plugins/featurePlugins/useChartCartesianAxis");
var _hooks = require("../hooks");
var _ChartsAxisHighlightPath = require("./ChartsAxisHighlightPath");
var _getSampledBandHighlight = require("./getSampledBandHighlight");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * @ignore - internal component.
 */
function ChartsXHighlight(props) {
  const {
    type,
    classes
  } = props;
  const {
    top,
    height
  } = (0, _hooks.useDrawingArea)();
  const store = (0, _useStore.useStore)();
  const axisXValues = store.use(_useChartCartesianAxis.selectorChartsHighlightXAxisValue);
  const xAxes = store.use(_useChartCartesianAxis.selectorChartXAxis);
  const bucketSizeByAxis = store.use(_useChartCartesianAxis.selectorChartHighlightBucketSize);
  if (axisXValues.length === 0) {
    return null;
  }
  return axisXValues.map(axisValue => {
    const {
      axisId,
      value
    } = axisValue;
    const xAxis = xAxes.axis[axisId];
    const xScale = xAxis.scale;
    const getXPosition = (0, _getValueToPositionMapper.getValueToPositionMapper)(xScale);
    const isXScaleOrdinal = type === 'band' && value !== null && (0, _scaleGuards.isOrdinalScale)(xScale);
    if (process.env.NODE_ENV !== 'production') {
      const isError = isXScaleOrdinal && xScale(value) === undefined;
      if (isError) {
        console.error([`MUI X Charts: The position value provided for the axis is not valid for the current scale.`, `This probably means something is wrong with the data passed to the chart.`, `The ChartsAxisHighlight component will not be displayed.`].join('\n'));
      }
    }
    let bandStart = 0;
    let bandSize = 0;
    if (isXScaleOrdinal) {
      ({
        bandStart,
        bandSize
      } = (0, _getSampledBandHighlight.getSampledBandHighlight)({
        scale: xScale,
        value,
        dataIndex: axisValue.dataIndex,
        data: xAxis.data,
        bucketSize: bucketSizeByAxis.get(axisId) ?? 1
      }));
    }
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
      children: [isXScaleOrdinal && xScale(value) !== undefined && /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChartsAxisHighlightPath.ChartsAxisHighlightPath, {
        d: `M ${bandStart} ${top} l ${bandSize} 0 l 0 ${height} l ${-bandSize} 0 Z`,
        className: classes.root,
        ownerState: {
          axisHighlight: 'band'
        }
      }), type === 'line' && value !== null && /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChartsAxisHighlightPath.ChartsAxisHighlightPath, {
        d: `M ${getXPosition(value)} ${top} L ${getXPosition(value)} ${top + height}`,
        className: classes.root,
        ownerState: {
          axisHighlight: 'line'
        }
      })]
    }, `${axisId}-${value}`);
  });
}
