// @mui/x-charts/BarChart/seriesConfig/bar/tooltipPosition.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.selectorTooltipItemPosition = void 0;
var _store = require("@mui/x-internals/store");
var _createGetBarDimensions = require("../../../internals/createGetBarDimensions");
var _useChartTooltip = require("../../../internals/plugins/featurePlugins/useChartTooltip/useChartTooltip.selectors");
var _useChartSeries = require("../../../internals/plugins/corePlugins/useChartSeries");
var _useChartCartesianAxisRendering = require("../../../internals/plugins/featurePlugins/useChartCartesianAxis/useChartCartesianAxisRendering.selectors");
const selectorTooltipItemPosition = exports.selectorTooltipItemPosition = (0, _store.createSelectorMemoized)(_useChartTooltip.selectorChartsTooltipItem, _useChartSeries.selectorChartSeriesProcessed, _useChartCartesianAxisRendering.selectorChartXAxis, _useChartCartesianAxisRendering.selectorChartYAxis, function selectorTooltipItemPosition(identifier, series, xAxes, yAxes,
/**
 * The preferred placement of the tooltip related to the element.
 * @default 'top'
 */
placement) {
  if (!identifier || identifier.type !== 'bar' || identifier.dataIndex === undefined) {
    return null;
  }
  const itemSeries = series.bar?.series[identifier.seriesId];
  if (!series.bar || !itemSeries) {
    return null;
  }
  const xAxis = xAxes.axis[itemSeries.xAxisId ?? xAxes.axisIds[0]];
  const yAxis = yAxes.axis[itemSeries.yAxisId ?? yAxes.axisIds[0]];
  if (!xAxis || !yAxis) {
    return null;
  }
  const groupIndex = series.bar.stackingGroups.findIndex(group => group.ids.includes(itemSeries.id));
  const dimensions = (0, _createGetBarDimensions.createGetBarDimensions)({
    verticalLayout: itemSeries.layout === 'vertical',
    xAxisConfig: xAxis,
    yAxisConfig: yAxis,
    series: itemSeries,
    numberOfGroups: series.bar.stackingGroups.length
  })(identifier.dataIndex, groupIndex);
  if (dimensions == null) {
    return null;
  }
  const {
    x,
    y,
    width,
    height
  } = dimensions;
  switch (placement) {
    case 'right':
      return {
        x: x + width,
        y: y + height / 2
      };
    case 'bottom':
      return {
        x: x + width / 2,
        y: y + height
      };
    case 'left':
      return {
        x,
        y: y + height / 2
      };
    case 'top':
    default:
      return {
        x: x + width / 2,
        y
      };
  }
});
